logagg_collector/collector.py

```diff
@@ -0,0 +1,926 @@
+                    self.log.exception('formatter_fn_not_found', fn=formatter)
+                    sys.exit(-1)
+                # Start a thread for every filepattern
+                log_f = dict(fpattern=fpattern,
+                                formatter=formatter, formatter_fn=formatter_fn)
+                log_key = (f['fpath'], fpattern, formatter)
+                if log_key not in self.log_reader_threads:
+
+                    self.log.info('starting_collect_log_files_thread', log_key=log_key)
+                    # There is no existing thread tracking this log file, start one.
+                    log_reader_thread = utils.start_daemon_thread(self._collect_log_files, (log_f,))
+                    self.log_reader_threads[log_key] = log_reader_thread
+
+        time.sleep(self.SCAN_FPATTERNS_INTERVAL)
+
+
+    @keeprunning(HEARTBEAT_RESTART_INTERVAL, on_error=utils.log_exception)
+    def _send_heartbeat(self, state):
+
+        # Sends continuous heartbeats to a seperate topic in nsq
+        if self.log_reader_threads:
+            files_tracked = [k for k in  self.log_reader_threads.keys()]
+        else:
+            files_tracked = ''
+
+        heartbeat_payload = {'namespace': self.NAMESPACE,
+                            'host': self.host,
+                            'port': self.port,
+                            'topic_name': self.master.topic_name,
+                            'files_tracked': files_tracked,
+                            'heartbeat_number': state.heartbeat_number,
+                            'timestamp': time.time(),
+                            }
+        self.nsq_sender_heartbeat.handle_heartbeat(heartbeat_payload)
+        state.heartbeat_number += 1
+        time.sleep(self.HEARTBEAT_RESTART_INTERVAL)
+
+
+    def collect(self):
+
+        # start tracking files and put formatted log lines into queue
+        state = AttrDict(files_tracked=list())
+        utils.start_daemon_thread(self._scan_fpatterns, (state,))
+
+        # start extracting formatted logs from queue and send to nsq
+        state = AttrDict(last_push_ts=time.time())
+        utils.start_daemon_thread(self._send_to_nsq, (state,))
+
+        # start sending heartbeat to "Hearbeat" topic
+        state = AttrDict(heartbeat_number=0)
+        self.log.info('init_heartbeat')
+        th_heartbeat = utils.start_daemon_thread(self._send_heartbeat, (state,))
+
+
+    def _fpath_in_trackfiles(self, fpath):
+        '''
+        Check the presence of fpath is in logaggfs trackfiles.txt
+        '''
+
+        # List of files in trackfiles.txt
+        with open(self.logaggfs.trackfiles, 'r') as f:
+            tf = f.readlines()
+
+        for path in tf:
+            if path[:-1] == fpath: return True
+        return False
+
+    def add_to_logaggfs_trackfile(self, fpath):
+        '''
+        Given a fpath add it to logaggfs trackfiles.txt via moving
+        '''
+        fd, tmpfile = tempfile.mkstemp()
```

Comment:
I think we need to close the fd because fd is the open file pointer, please do check if fd is file pointer or not.